File: Utilities/Matlab/scripts/flame_species.m
%------------------------------------------------
% C Weinschenk
% 10-2011
% Combine outputs from Methane_flame_lumped and
% Methane_flame_primitive into 1 file
%------------------------------------------------

%-----------------------
% Write directory
%-----------------------

FDS_Output_Files = '../../Verification/Species/';

%----------------------
% Import files
%----------------------

if ~exist([FDS_Output_Files,'methane_flame_primitive_devc.csv'])
    display(['Error: File ',[FDS_Output_Files,'methane_flame_primitive_devc.csv'],' does not exist. Skipping case.'])
    return
end

if ~exist([FDS_Output_Files,'methane_flame_lumped_devc.csv'])
    display(['Error: File ',[FDS_Output_Files,'methane_flame_lumped_devc.csv'],' does not exist. Skipping case.'])
    return
end

if ~exist([FDS_Output_Files,'methane_flame_lumped_fuel_devc.csv'])
    display(['Error: File ',[FDS_Output_Files,'methane_flame_lumped_fuel_devc.csv'],' does not exist. Skipping case.'])
    return
end

if ~exist([FDS_Output_Files,'methane_flame_lumped_ox_devc.csv'])
    display(['Error: File ',[FDS_Output_Files,'methane_flame_lumped_ox_devc.csv'],' does not exist. Skipping case.'])
    return
end

if ~exist([FDS_Output_Files,'methane_flame_primitive_2_devc.csv'])
    display(['Error: File ',[FDS_Output_Files,'methane_flame_primitive_2_devc.csv'],' does not exist. Skipping case.'])
    return
end


primitive_struct=importdata([FDS_Output_Files,'methane_flame_primitive_devc.csv']);
primitive_data=primitive_struct.data;
primitive_text=primitive_struct.textdata;

lumped_struct=importdata([FDS_Output_Files,'methane_flame_lumped_devc.csv']);
lumped_data=lumped_struct.data;
lumped_text=lumped_struct.textdata;

header1=cat(2,primitive_text(1),lumped_text(1,2:4));
header2=cat(2,primitive_text(2,:),lumped_text(2,2:4));
data=cat(2,primitive_data,lumped_data(:,2:4));

primitive_struct_2=importdata([FDS_Output_Files,'methane_flame_primitive_2_devc.csv']);
primitive_data_2=primitive_struct_2.data;
primitive_text_2=primitive_struct_2.textdata;

lumped_struct_f=importdata([FDS_Output_Files,'methane_flame_lumped_fuel_devc.csv']);
lumped_data_f=lumped_struct_f.data;
lumped_text_f=lumped_struct_f.textdata;

lumped_struct_ox=importdata([FDS_Output_Files,'methane_flame_lumped_ox_devc.csv']);
lumped_data_ox=lumped_struct_ox.data;
lumped_text_ox=lumped_struct_ox.textdata;

header1b=cat(2,primitive_text_2(1),lumped_text_f(1,2:3),lumped_text_ox(1,2:3));
header2b=cat(2,primitive_text_2(2,:),lumped_text_f(2,2:3),lumped_text_ox(2,2:3));
datab=cat(2,primitive_data_2,lumped_data_f(:,2:3),lumped_data_ox(:,2:3));

%-----------------------
% Write new files
%-----------------------

fid = fopen([FDS_Output_Files,'methane_flame_lumpedprimitive.csv'],'wt','n');

fprintf(fid,'%s,%s,%s,%s,%s,%s,%s\n','s','kg','kg','kg','kg','kg','kg');
fprintf(fid,'%s,%s,%s,%s,%s,%s,%s\n',header2{1,:});
for i=1:length(primitive_data)
    fprintf(fid,'%f,%f,%f,%f,%f,%f,%f\n',data(i,:));
end
fclose(fid);

fid = fopen([FDS_Output_Files,'methane_flame_multilumped.csv'],'wt','n');

fprintf(fid,'%s,%s,%s,%s,%s,%s,%s\n','s','kg','kg','kg','kg','kg','kg');
fprintf(fid,'%s,%s,%s,%s,%s,%s,%s\n',header2b{1,:});
for i=1:length(primitive_data_2)
    fprintf(fid,'%f,%f,%f,%f,%f,%f,%f\n',datab(i,:));
end
fclose(fid);

File: Utilities/Python/scripts/flame_species.py
# C Weinschenk
# flame_species.m
# 10-2011
# Combine outputs from Methane_flame_lumped and
# Methane_flame_primitive into 1 file
# ------------------------------------------------
#
# Converted by Floyd
# 10/16/2025

import pandas as pd
import numpy as np
import os
import csv
from typing import List, Tuple, Optional


# Write directory
outdir = '../../Verification/Species/'
filename = [
    'methane_flame_primitive_devc.csv',
    'methane_flame_primitive_2_devc.csv',
    'methane_flame_lumped_devc.csv',
    'methane_flame_lumped_fuel_devc.csv',
    'methane_flame_lumped_ox_devc.csv',
]

skip_case = False

for i in range(len(filename)):
   name = outdir+filename[i]
   if not os.path.exists(name):
      skip_case = True
      print('Error: File ', filename[i], ' does not exist. Skipping case.')

if skip_case: quit()

h_mf_p = pd.read_csv(outdir+filename[0],skiprows=1,nrows=1,header=None)
h_mf_p_2 = pd.read_csv(outdir+filename[1],skiprows=1,nrows=1,header=None)
h_mf_l = pd.read_csv(outdir+filename[2],skiprows=1,nrows=1,header=None)
h_mf_l_f = pd.read_csv(outdir+filename[3],skiprows=1,nrows=1,header=None)
h_mf_l_o = pd.read_csv(outdir+filename[4],skiprows=1,nrows=1,header=None)

mf_p = pd.read_csv(outdir+filename[0],skiprows=2,header=None)
mf_p_2 = pd.read_csv(outdir+filename[1],skiprows=2,header=None)
mf_l = pd.read_csv(outdir+filename[2],skiprows=2,header=None)
mf_l_f = pd.read_csv(outdir+filename[3],skiprows=2,header=None)
mf_l_o = pd.read_csv(outdir+filename[4],skiprows=2,header=None)

header_lp = tuple(h_mf_p.iloc[0,:]) + tuple(h_mf_l.iloc[0,1:])

print(header_lp)

data_combined_lp = pd.concat([mf_p, mf_l.iloc[:, 1:4]], axis=1)
data_combined_lp.columns = header_lp

print(data_combined_lp)

header_ml =  tuple(h_mf_p_2.iloc[0,:]) +  tuple(h_mf_l_f.iloc[0,1:3]) +  tuple(h_mf_l_o.iloc[0,1:3])
print(header_ml)
data_combined_ml = pd.concat([mf_p_2, mf_l_f.iloc[:, 1:3], mf_l_o.iloc[:, 1:3]], axis=1)
data_combined_ml.columns = header_ml

print(data_combined_ml)

output_file_1 = outdir+'methane_flame_lumpedprimitive2.csv'

# The MATLAB script uses a hardcoded units line for the final output.
hardcoded_units = ['s', 'kg', 'kg', 'kg', 'kg', 'kg', 'kg']

# Create a temporary DataFrame for the two header rows
header_df_1 = pd.DataFrame([hardcoded_units, header_lp], columns=data_combined_lp.columns)

# Concatenate headers and data
df_out_1 = pd.concat([header_df_1, data_combined_lp]).reset_index(drop=True)

# Write to CSV without the pandas index or automatically generated header
df_out_1.to_csv(output_file_1, header=False, index=False)
print(f"Successfully wrote combined data to {output_file_1}")

output_file_2 = outdir+'methane_flame_multilumped2.csv'

# Create a temporary DataFrame for the two header rows
header_df_2 = pd.DataFrame([hardcoded_units, header_ml], columns=data_combined_ml.columns)

# Concatenate headers and data
df_out_2 = pd.concat([header_df_2, data_combined_ml]).reset_index(drop=True)

# Write to CSV without the pandas index or automatically generated header
df_out_2.to_csv(output_file_2, header=False, index=False)
print(f"Successfully wrote multi-lumped data to {output_file_2}")
